docs/rfcs/0001-integration-with-dune.md

```diff
@@ -0,0 +1,169 @@
+    (modes melange)
+    (melange (module es6))
+)
+```
+- add an option to choose which suffix will be added to JavaScript files
+(`.bs.js` is the default), e.g.:
+```dune
+(executable
+    (name foo)
+    (modes melange)
+    (melange (suffix bs.js))
+)
+```
+- add implicit handling targets with extension `.bs.js`, which will generate the
+necessary build rules behind the scenes, same as it is done currently for
+`bc.js` files in Js_of_ocaml.
+
+### `.js` targets
+
+One of the main challenges of Melange integration is that originally `bsc`
+command would generate both `cmo` and `js` targets with same command, but they
+need to be placed in separate directories. This does not play well with Dune
+invariants, that require all targets of a single rule
+[to be placed in same directory](https://github.com/ocaml/dune/blob/0490fd413b81cca473ef5e4fb0a6c876e000e953/src/dune_engine/rule.ml#L96-L110).
+
+After [some recent changes](https://github.com/melange-re/melange/pull/199),
+[@EduardoRFS](https://github.com/EduardoRFS) unlocked the possibility to
+generate `js` files from `cmo` files, which allows to define `cmo` and `js`
+artifacts with separate Dune rules. This allows to generate just `cmo` files
+first:
+
+```bash
+ocamlc -o output.cmo -bs-stop-after-cmo -impl output.ml
+```
+
+And then generate `js` files:
+
+```bash
+ocamlc -o output.js output.cmo
+```
+
+One remaining question is how to resolve the `import` / `require` statements in
+resulting JavaScript files, as they require knowledge about the relative paths
+to each one of the dependent modules that a module depends on.
+
+Right now, Melange [checks if the `js` file exist in the relative path](https://github.com/jchavarri/melange/blob/6524091f126ec720f7fb59873d18bdff9b001be2/jscomp/core/js_name_of_module_id.ml#L175) to get absolute paths to dependent modules, so in
+order to solve the issue we could include both `cmo` and `js` paths when calling
+`ocamlc`:
+```bash
+ocamlc -I where_your_cm_stuff_is -I where_your_js_is exe.cmo
+```
+
+But there might better approaches, like passing all load paths to dependencies
+from the command line argument.
+
+# Drawbacks
+
+From users perspective, integration with Dune means that in-source compilation
+would no longer be supported. This has upsides like more easily shareable
+artifacts or caching, and removes the chances of accidentally committing build
+artifacts under source control. But it can also make harder to configure certain
+projects that require `js` assets to be co-locating with other assets like `css`
+or images. This can be solved by using Dune copy rules.
+
+Another downside is that this integration will add maintenance burden to Dune,
+as it adds more modes, and expands the amount of use cases that need to be
+supported. This is partially mitigated by leveraging cross-compilation.
+
+It similarly adds more maintenance costs to Melange, which would need to support
+both `bsconfig.json` workflow, as well as `dune` workflow.
```

Comment:
I would vote for slowly deprecating bsb, having a proper tool to migrate from bsconfig to dune, it shouldn't be hard just a bit of `(wrapped false)` and then we keep this in maintenance for some well known time, like 6-12 months.

Maybe moving bsb fully to userspace it's a first step.